### PR TITLE
MouseTracker Overhaul 2020

### PR DESCRIPTION
**Pull request by msalsbery**
## MouseTracker Overhaul 2020
Finally finishing up this major MouseTracker overhaul! This should fix many of the mouse/touch/pen device issues we've seen over the past several years! Also added improved capabilities for controlling event traversal through the DOM tree, and support for using OpenSeadragon viewers in iframe and shadow DOM should be much improved!

### Main Goals
* Use native pointer capture whenever available. This is less prone to issues and makes handling native events in shadow DOM and frames much simpler
* Add a hook handler for pre-processing native events and allowing finer-grained control over the disposition of the events (bubbling, default handling). This should make element layering above/below OpenSeadragon viewers much simpler
* Adjusting the rest of the library to use the new event processing properly, eliminating tweaks/hacks that were necessary before
* Properly maintain/clean up internal gesture tracking objects

### Change Log
* Improved browser sniffing - detect EDGE and CHROMEEDGE browsers
* Improved DOM event model feature detection
* Added support for options parameter on addEvent()/removeEvent (to support passive option) (#1669)
* Added OpenSeadragon.eventIsCanceled() function for defaultPrevented detection on DOM events
* MouseTracker: better PointerEvent model detection - removed use of deprecated window.navigator.pointerEnabled
* MouseTracker: added overHandler/outHandler options for handling corresponding pointerover/pointerout events
* MouseTracker: changed enterHandler/leaveHandler to use DOM pointerenter/pointerleave events instead of simulating using pointerover/pointerout
* All internal uses of MouseTracker use pointerenter/pointerleave events instead of pointerover/pointerout events for more consistent pointer tracking
* DEPRECATION: MouseTracker exitHandler deprecated for name change to leaveHandler for consistency with DOM event names
* Fixed bug in Button class where two MouseTracker event handlers used an invalid "this" causing issues in some browsers
* Added pointerType property to Viewer container-enter, container-exit, canvas-drag, canvas-drag-end, canvas-pinch events
* MouseTracker: Fire dragEndHandler event even if release point same as initial contact point (#1459)
* MouseTracker: Pointer capture implemented with capture APIs where available. Only fallback to emulated capture on extremely old browsers
* MouseTracker: Added preProcessEventHandler option to allow MouseTracker instances to control bubbling and default behavior of events on their associated element
* MouseTracker: Improved handling of canceled events (#1728)
* MouseTracker: Improved releasing of tracked pointers on destroy()/stopTracking() (#1346)
* Updated Viewer, Button, Drawer, Navigator, ReferenceStrip DOM for proper DOM event handling
* Added OpenSeadragon.setElementPointerEventsNone() for setting pointer-events:'none' on DOM elements
* MouseTracker: added contextMenuHandler option for handling contextmenu events
* Viewer: added a canvas-contextmenu event
* MouseTracker: Per #1863, dropped support for Internet Explorer < 11

### CODE REVIEW

**Comment by msalsbery** [comment]
## MouseTracker Overhaul 2020 Preview
Finally finishing up this major MouseTracker overhaul! This should fix many of the mouse/touch/pen device issues we've seen over the past several years! Also added improved capabilities for controlling event traversal through the DOM tree, and support for using OpenSeadragon viewers in iframe and shadow DOM should be much improved!

### Beta Testers Needed!
As I clean up a punch list of small issues I found along the way, if anyone would like to test it out it would help immensely! Links below...

### Demo Sites
[OpenSeadragon Imaging ImagingHelper Demo](https://openseadragon-imaging.github.io/#/imaginghelper)
[OpenSeadragon Imaging ConsoleHook Demo](https://openseadragon-imaging.github.io/#/consolehook)
[Legacy OpenSeadragon Imaging ImagingHelper Demo](https://openseadragon-imaging.github.io/old-demo/index.html)

### Preview Builds
[openseadragon.js](https://openseadragon-imaging.github.io/old-demo/lib/openseadragon.js)
[openseadragon.js.map](https://openseadragon-imaging.github.io/old-demo/lib/openseadragon.js.map)
[openseadragon.min.js](https://openseadragon-imaging.github.io/old-demo/lib/openseadragon.min.js)
[openseadragon.min.js.map](https://openseadragon-imaging.github.io/old-demo/lib/openseadragon.min.js.map)

**Comment by nagame-net** [comment]
I just tested in my web application (which uses frames), and the rendering is perfect. It works much better than before. Thanks a lot for the patch!

**Comment by nagame-net** [comment]
Support on Firefox is a bit better than on Webkit. In Chrome, the zoom movement stops when the mouse is outside the frame area. 
With Firefox the movement continues even if the mouse is outside the frame area.

**Comment by msalsbery** [comment]
> In Chrome, the zoom movement stops when the mouse is outside the frame area.
> With Firefox the movement continues even if the mouse is outside the frame area.

Chrome running on what platform?

Thank you for testing!! Really appreciate the feedback!

**Comment by msalsbery** [comment]
> Support on Firefox is a bit better than on Webkit.

How so? I started the process of submitting a Webkit bug two days ago, wondering if it's related. It's possible to work around, and I tried to get at least the built-in OpenSeadragon event handling to work around it, but I only have so many devices/browsers available to test on (which makes feedback like yours invaluable!)

**Comment by nagame-net** [comment]
I am using Windows 10. Note I use frames (not an iframe). I want to stop using frames, but it's a fairly large application, requiring a lot of code changes. In your demo, the bug does not exist, the movement of the zoom works well.

**Comment by nagame-net** [comment]
> Edit: Also, does it recover - in other words, can you still do stuff in the frame if you release outside of the frame or does it lock up

Absolutely. In Chrome, if I release the mouse button outside the frame area, it locks on the zoom. I must then close the zoom and reopen it

(sorry for my poor English :-)

**Comment by msalsbery** [comment]
> In Chrome, if I release the mouse button outside the frame area, it locks on the zoom. I must then close the zoom and reopen it

Just to clarify...this is Chrome on Windows 10, correct? And the new patched code stops zooming when button released outside the frame the button press occurred in?

File: changelog.txt
Comment (iangilman) [reply changelog.txt:19]: Looking through the code it appears that the exitHandler is entirely removed, not just deprecated. I would call that a breaking change. We should be able to just deprecate it, though… Continue to provide the hook for it, but don't document it (or document it as deprecated). If the user attaches to it, put a warning in the console (but allow it to continue to work).

File: changelog.txt
Comment (iangilman) [reply changelog.txt:11]: Feel free to officially bump up the in progress number at the top to 2.5.0, BTW :-)

File: changelog.txt
Comment (msalsbery) [reply changelog.txt:19]: > Looking through the code it appears that the exitHandler is entirely removed, not just deprecated. I would call that a breaking change. We should be able to just deprecate it, though… Continue to provide the hook for it, but don't document it (or document it as deprecated). If the user attaches to it, put a warning in the console (but allow it to continue to work).

It's still there, documented as deprecated, and called if a handler provided.

File: changelog.txt
Comment (msalsbery) [reply changelog.txt:11]: > We are actually putting the pull request number, not the issue number, on items in the change log, and including the username of the patch author. I know it may feel a little silly, but I think it's worth listing all of these changeling items separately like you're doing, but including the same PR number and author at the end of each one.

Changelog is next on my punch list since you mentioned the format in the last PR that landed  ;) ...fixing

File: src/mousetracker.js
Comment (msalsbery) [inline src/mousetracker.js:3454]: @iangilman  Here's where the deprecated exitHandler is handled. Pondering if I should use an else-if so only one handler is called...and maybe emit a warning to the console if it's exitHandler?

**Comment by INPUTsys-Chris** [comment]
Hi @msalsbery 

Great patch! Firefox panning now works!

At least in shadow dom it seems that click and dblClick dont't work anymore. We add MouseTrackers with dblClick and click in an addOnceHandler('open'

Should I try to develop a SSCCE with or without shadow dom?

Anyway, thank you for that major overhaul!

**Comment by msalsbery** [comment]
@INPUTsys-Chris Thank you for taking a look!

Does drag work in shadow DOM while click/dblclick don't work?

**Comment by INPUTsys-Chris** [comment]
Panning/dragging the image did work.

Thank you!

**Comment by msalsbery** [comment]
> Panning/dragging the image did work.
> 
> 
> 
> Thank you!

Cool thanks!

In that case, click/dblclick should work as they're generated from the same code path. There may be an issue with using them addOnce however. If you have a little code example of how you're doing that I can look into it.

Thanks much for the feedback, I'm very pleased to hear it's working in shadow DOM.

**Comment by msalsbery** [comment]
React users may like this info...

[React v17.0 Release Candidate: No New Features](https://reactjs.org/blog/2020/08/10/react-v17-rc.html)

No new features, perhaps, but the new event handling is a great step, and a lovely, timely change coinciding with this patch.

Yay!

File: src/mousetracker.js
Comment (iangilman) [reply src/mousetracker.js:3496]: Thank you for pointing it out here! I think it's probably fine if both handlers are called (typically if you have an exit handler you wouldn't have a leave handler anyway). I think it would be good to console.warn if it's the exit handler, though.

**Review by iangilman — COMMENTED** [review]
Wow, this patch is a beast! Fantastic to have all of these issues fixed!

I can't say I fully understand every detail of the patch, but I've looked through the whole thing and it looks pretty sane. I like the new organization; seems very methodical and precise. Keeping things unified where they can be unified, but paying attention to differences where needed.

I can't help but notice that the resulting mousetracker.js is almost 1000 lines longer! I wonder if there's any way to slim that down (without compromising clarity and ease of future development.

A bunch of the new size is commented out lines from previous hacks that hopefully won't be needed, but we're not sure yet. I think that's the correct approach there, and I'm not concerned about them from a size standpoint, since they'll get removed in the minification process (and they don't contribute to the complexity of the code at all).

Anyway, something to think about. Maybe there are ways we can unify some of the handlers (though I wouldn't want that if it meant making them more complicated/confusing with overlapping meaning).

We'll want to remember to actually close all of the issues that are fixed with this patch when it lands. You can automate that by adding "fixes #1824, fixes #1728" etc. in the top comment/description of this PR.

Just to confirm, what's the status on allowing click events and such up through the viewer so they can be handled by regular JavaScript in overlays and whatnot? Does this patch take care of that?

Another random thought: I see you have code for mouseover/mouseout handling… Do we need that? I think mouseenter/mouseleave is preferable.

So, besides additional testing, what would you say is left to do on this patch?

So exciting to have this coming together! 🎉🎉🎉

**Comment by msalsbery** [comment]
> I can't help but notice that the resulting mousetracker.js is almost 1000 lines longer! I wonder if there's any way to slim that down (without compromising clarity and ease of future development.

I would really like to remove the THIS[this.hash] stuff and just use straight-forward privates named with underscore and documented appropriately. It's cool that it hides private stuff but readability-wise and more importantly managing `this` can be confusing and/or problematic, I did want to ask you your opinion about that first though, which I'm asking now ;-) ...

> A bunch of the new size is commented out lines from previous hacks that hopefully won't be needed, but we're not sure yet. I think that's the correct approach there, and I'm not concerned about them from a size standpoint, since they'll get removed in the minification process (and they don't contribute to the complexity of the code at all).

That was my thinking, especially before I got any outside eyes on it - in case I had to bring any of that back in. I'm pretty comfortable with removing those at this point, however. If there's still issues anywhere then fixes will have to be re-written anyway

> We'll want to remember to actually close all of the issues that are fixed with this patch when it lands. You can automate that by adding "fixes #1824, fixes #1728" etc. in the top comment/description of this PR.

Cool thanks! Didn't know it could be automated! I put the links so I would remember where the issues are and to update the fixed issues

> Just to confirm, what's the status on allowing click events and such up through the viewer so they can be handled by regular JavaScript in overlays and whatnot? Does this patch take care of that?

Yes, there's now a hook event to give an opportunity to set flags to control the disposition of DOM events. I didn't do anything OpenSeadragon=specific...I actually made it work and read more like DOM events so it would be less confusing to users. Users still need to understand DOM event processing to do advanced stuff, but it's the same as the spec, nothing new, strange, or fancy to add confusion
 
> Another random thought: I see you have code for mouseover/mouseout handling… Do we need that? I think mouseenter/mouseleave is preferable.

Indeed, enter/leave is preferable in the majority of cases. over/out are now exposed there just for consistency but no longer used internally.

> So, besides additional testing, what would you say is left to do on this patch?

Just testing and fixes you've already commented on (in this latest comment, the contextmenu thing, and adding a console warning for the deprecated exitHandler)

> So exciting to have this coming together! 🎉🎉🎉

YES! I'm excited...not only does it work much better, I needed the new features that were holding up development of my plugin suite!

**Comment by msalsbery** [comment]
@iangilman 

OH! Also before this lands I wanted to remove the IE<11 code from MouseTracker...I'll get the rest of the library in a separate patch

**Comment by iangilman** [comment]
> I would really like to remove the THIS[this.hash] stuff and just use straight-forward privates named with underscore and documented appropriately. It's cool that it hides private stuff but readability-wise and more importantly managing this can be confusing and/or problematic, I did want to ask you your opinion about that first though, which I'm asking now ;-) ...

Yes, please remove it! I don't even think it's good that it's hiding private stuff… Makes it really hard to debug. 

> That was my thinking, especially before I got any outside eyes on it - in case I had to bring any of that back in. I'm pretty comfortable with removing those at this point, however. If there's still issues anywhere then fixes will have to be re-written anyway

Cool. I'm fine either way (before or after this patch lands).

> Yes, there's now a hook event to give an opportunity to set flags to control the disposition of DOM events. I didn't do anything OpenSeadragon=specific...I actually made it work and read more like DOM events so it would be less confusing to users. Users still need to understand DOM event processing to do advanced stuff, but it's the same as the spec, nothing new, strange, or fancy to add confusion

When you say there's a hook, does that mean the developer needs to do something special? Or, by default, is it already set up so they can just do normal event handling? In other words, if there is an overlay with a link in it, we would just work by default? Or do people still need to do hacks like https://codepen.io/iangilman/pen/mEkKaY? Or some new thing?

> Indeed, enter/leave is preferable in the majority of cases. over/out are now exposed there just for consistency but no longer used internally.

Is it something we are keeping for backward compatibility? If not, and we're not using it, maybe we should get rid of it?

**Comment by msalsbery** [comment]
> Yes, please remove it! I don't even think it's good that it's hiding private stuff… Makes it really hard to debug.

Yay, thank goodness you feel that way as well!

> Cool. I'm fine either way (before or after this patch lands).

Right on, I'll remove the commented out code before this lands

> When you say there's a hook, does that mean the developer needs to do something special? Or, by default, is it already set up so they can just do normal event handling? In other words, if there is an overlay with a link in it, we would just work by default? Or do people still need to do hacks like https://codepen.io/iangilman/pen/mEkKaY? Or some new thing?

That doesn't appear to be a hack to me...it looks like a valid use of event delegation (handling events at an ancestor element and delegating to descendants. That specific example is possibly the simplest, cleanest solution if it does what you want it to do. 

Anyway, by "hook", I just mean a callback that, if provided, provides a chance to look at the DOM event and possibly alter whether it 
* gets canceled (default user agent action prevented)
* propagates (bubbles) up the DOM tree 
* gets used to form a gesture
The biggest issue corrected by the patch is not stopping propagation of DOM events. In a perfect world, there shouldn't be a need to stop events from bubbling...we used to stop propagation of every event we used, so ancestor elements couldn't handle them if they needed to. But, there's nothing built in to the spec to notify ancestors that "hey, I've already handled the event so you just ignore it when it bubbles up to you", so sometimes, depending on the ancestor implementation, it's necessary or desired to stop propagation.

A typical way to notify ancestor elements that an event has been handled so it can be ignored when it bubbles is to use the event's preventDefault() method to set the defaultPrevented flag on the event. That's a pretty common way of handling, and should lessen the issues when stacking elements that are part of frameworks that do their own event handling on top of our viewers.

The only thing I've done here is removed the clamped-down, no-propagation event handling so it's as close to what a user would have to do if handling events directly. No magic. Users do need to understand how events in the DOM work, and with this patch the event handling is much more standard (less unknowns or quirks) and now there's a way to control event propagation and default handling for advanced scenarios which didn't exist before. MouseTracker is much less of a black box now.

Hopefully that makes sense! High on my TODO list is an entire new page on the OpenSeadragon site explaining basic event handling with MouseTracker, as well as probably a wiki article on the Github page with a deeper dive into DOM events and how our event handling fits in. I really don't want to have to explain it every time a new user asks how can I drag my annotations without the viewer also panning ;)

> Is it something we are keeping for backward compatibility? If not, and we're not using it, maybe we should get rid of it?

We're not using those events internally, but if a user wants to see those events on an element a MouseTracker is attached to, it seems like it would be cleaner for the user to simply provide MouseTracker callbacks for them instead of having to attach handlers to the element directly outside of the MouseTracker that's already attached to the element. The extra code needed in MouseTracker to do this is minimal, while the user only has to provide a callback and let MouseTracker handle the adding and removing of DOM event handlers.

That does remind me about this, however. which I wanted to run by you...

Something I want to do, but was going to hold off for a next patch, is to have MouseTracker only add internal event handlers for the events needed by MouseTracker to be able to create the gesture event callbacks the user subscribed to. Currently (and it also passed through the last big overhaul 5-6 years ago) we subscribe to every event that we could possibly use, so even if a user just attached a MouseTracker to an element to see enter/leave events, the MouseTracker is still handling touches, presses, releases, etc. Definitely not ideal ;-)  Now that you bring up the questions I've discussed above, it feels to me like this may be a good time to rework that as well, especially with the green light to clean up the hidden hash table stuff!

**Comment by msalsbery** [comment]
> > When you say there's a hook, does that mean the developer needs to do something special? Or, by default, is it already set up so they can just do normal event handling? In other words, if there is an overlay with a link in it, we would just work by default? Or do people still need to do hacks like https://codepen.io/iangilman/pen/mEkKaY? Or some new thing?
> 
> That doesn't appear to be a hack to me...it looks like a valid use of event delegation (handling events at an ancestor element and delegating to descendants. That specific example is possibly the simplest, cleanest solution if it does what you want it to do.

@iangilman  I feel I need to expand on my comment regarding the codepen...

I presume the "hack" was to get the anchor link to work when there's an OpenSeadragon viewer above the anchor in the DOM tree? I can see that happening, since the MouseTracker on the viewer's canvas (the DIV that catches all the pointer events to provide panning, clicks, etc.) prevents pointer event propagation to the document level where default anchor processing occurs.

This patch now lets those events propagate, but there's still an issue for scenarios like in the codepen: The viewer's MouseTracker is handling down, move, and up events for panning. Because of this, preventDefault() is being called, as it should be, on the down events so the default anchor click handling is STILL not going to happen even though the events are bubbling to the document level.

This is not a MouseTracker issue - this is a DOM event handling issue. Remove MouseTrackers from the equation and make an element that handles pointer down, move, up events to pan something in it. Now add a child anchor element. User clicks the anchor - the down event needs to bubble to the document level for default processing of the link, but the parent of the anchor is also handling the down event. As I mentioned in my previous post, there's no built-in mechanism for a child to notify its parent that it's handled the event except for calling preventDefault(). But this is an anchor element, a standard HTML element. It doesn't handle the event to call preventDefault()....event processing is done at the document level. It's possible one could find a browser where anchor elements do handle the events and call preventDefault(), but it's not in the spec and can't be relied upon or expected in any way. This makes sense, I suppose, because who would need to process any events "under" an anchor element - clickable elements on top of each other? Only scenarios like OpenSeadragon ;-)

So that's why when I look at the codepen, I see a reasonable solution to a complex issue that can't be handled generically. Any elements that allow pointer interaction used as a child of another element that provides pointer interaction (like an OpenSeadragon viewer, for instance) are going to face this issue. 

This patch removes the hidden behaviors to at least give the user a chance to implement solutions for these scenarios instead of just swallowing the events, but making it so hacks like in that codepen aren't necessary isn't really possible.

It's the infinite number of possibilities like this that makes these MouseTracker patches take so long....trying to imagine all the common scenarios and make them possible to handle in a generic way ;)

**Comment by iangilman** [comment]
> Hopefully that makes sense! High on my TODO list is an entire new page on the OpenSeadragon site explaining basic event handling with MouseTracker, as well as probably a wiki article on the Github page with a deeper dive into DOM events and how our event handling fits in.

Yes! That would be fantastic.

> We're not using those events internally, but if a user wants to see those events on an element a MouseTracker is attached to, it seems like it would be cleaner for the user to simply provide MouseTracker callbacks for them instead of having to attach handlers to the element directly. 

Fair enough. Might as well leave it in for completeness.

Keep in mind, though, that as awesome as the MouseTracker is, it's unfamiliar to most people, and they already have an event handling technique they prefer (whether it's jQuery or vanilla JavaScript or whatever). So for them, using what they are already used to is the preferred path. That said, if they are already using the MouseTracker for other things, and they also needed over/out support, then I could see it being nice to keep the handlers together.

> have MouseTracker only add internal event handlers for the events needed by MouseTracker to be able to create the gesture event callbacks the user subscribed to.

Sounds good! Of course this is already a big patch, so if that's another good sized chunk of work, it could make sense to make it a follow-up. Whatever you think is best.

> I presume the "hack" was to get the anchor link to work when there's an OpenSeadragon viewer above the anchor in the DOM tree?

Exactly. There must be some way we can allow the link to process its own events and not have to manually open a window ourselves. If we take OpenSeadragon out of the equation and just have some containing element that does something like panning based on mousedown, etc., I believe all we would need to do is add this code:

```
$('.overlay1').on('mousedown', function(event) { event.stopPropagation(); });
```

… And that would be enough to make the overlay not engage the panning, and the link to handle its own click. At least that's what I've done and other projects with this sort of scenario.

To be honest I haven't tried that lately on OpenSeadragon (or with the new patch), so I don't know for sure that it doesn't work, but I seem to recall doesn't.

I understand that the panning complicates things, and we're not going to be able to avoid needing to control the events, but having to manually replicate the link functionality seems unfortunate. Can you give some thought as to how we might make it possible for the link to handle its own event, and what code the user would need to use?

**Comment by msalsbery** [comment]
> > We're not using those events internally, but if a user wants to see those events on an element a MouseTracker is attached to, it seems like it would be cleaner for the user to simply provide MouseTracker callbacks for them instead of having to attach handlers to the element directly.
> 
> Fair enough. Might as well leave it in for completeness.

I keep flip-flopping in my head on this...I think may remove it lol

> Keep in mind, though, that as awesome as the MouseTracker is, it's unfamiliar to most people, and they already have an event handling technique they prefer (whether it's jQuery or vanilla JavaScript or whatever). So for them, using what they are already used to is the preferred path. That said, if they are already using the MouseTracker for other things, and they also needed over/out support, then I could see it being nice to keep the handlers together.

Yeah... Hopefully a page on the site will advertise it so more people are familiar. There's no magic there, but if one is already using OpenSeadragon and needs common pointer event handling, and even more so gestures like drag and click, using MouseTracker eliminates having to write a LOT of code, especially if robust coverage of the different legacy/modern APIs is desired

> > have MouseTracker only add internal event handlers for the events needed by MouseTracker to be able to create the gesture event callbacks the user subscribed to.
> 
> Sounds good! Of course this is already a big patch, so if that's another good sized chunk of work, it could make sense to make it a follow-up. Whatever you think is best.

I think I'll defer this to a subsequent patch...it'll be easier for development for me to have a clean diff to work with in my self-review and testing before submission

> > I presume the "hack" was to get the anchor link to work when there's an OpenSeadragon viewer above the anchor in the DOM tree?
> 
> Exactly. There must be some way we can allow the link to process its own events and not have to manually open a window ourselves. If we take OpenSeadragon out of the equation and just have some containing element that does something like panning based on mousedown, etc., I believe all we would need to do is add this code:
> 
> ```
> $('.overlay1').on('mousedown', function(event) { event.stopPropagation(); });
> ```
> 
> … And that would be enough to make the overlay not engage the panning, and the link to handle its own click. At least that's what I've done and other projects with this sort of scenario.

> To be honest I haven't tried that lately on OpenSeadragon (or with the new patch), so I don't know for sure that it doesn't work, but I seem to recall doesn't.
> 
> I understand that the panning complicates things, and we're not going to be able to avoid needing to control the events, but having to manually replicate the link functionality seems unfortunate. Can you give some thought as to how we might make it possible for the link to handle its own event, and what code the user would need to use?

I'll put together some examples of solutions (here, before we land this) which may make it more clear than my rambling text ;)  

Your example at first glance I would not expect to work because if the down event isn't allowed to bubble to the document level then the default handling for the link won't happen. But again, this is not unique to MouseTracker... if you just had an ancestor element above the link that handles down events then you'd still have the same issue, which is deciding to handle the event or let it bubble. It's a universal issue (and understandable why the API doesn't make it simpler because typically an anchor element wouldn't have an ancestor that cares about pointer events). One possible solution, with this patch, one can add a Viewer.preProcessEvent handler to add code like your previous example that checked if the target is an anchor element and if so then let it pass through by simply setting the preventGesture property to true.

Anyway, I'll put together examples for you here (which I'll reuse when I add new info to site-build site) since I'm once again trying to describe it in rambling text ;)

**Comment by msalsbery** [comment]
BUMP! Testers needed! ;-) I want to get this thing DONE!

**Comment by msalsbery** [comment]
@iangilman 

This is pretty close to being able to land!

Here's a copy of your Codepen (link on an overlay) using the new MouseTracker preProcessEvent handling to control event propagation/disposition: https://codepen.io/msalsbery/pen/dyOeXqO
Note it eliminates the need to manually handle clicks on the link!

**Comment by msalsbery** [comment]
@iangilman  Update: Removing old IE support which will allow me to eliminate a bunch of redundant code in mousetracker.js..._THEN_ I'll be ready to land this thing!

**Comment by msalsbery** [comment]
@iangilman Update:

Ready for review!

I removed a lot of redundant code, as well as old IE version support code, from MouseTracker. The added lines number should be much more reasonable now ;)

**Review by iangilman — APPROVED** [review]
@msalsbery exciting!

Thank you for the updated CodePen… Exciting to see us now able to pass the event through directly! It might be nice to add some sugar for this scenario at some point, but just being able to do it at all is a big step :-)

Man, this PR is a behemoth! I looked it over, and it seems reasonable from what I can tell. I think at this point we should merge it and have folks test! What do you think?

**Comment by msalsbery** [comment]
> Man, this PR is a behemoth! I looked it over, and it seems reasonable from what I can tell. I think at this point we should merge it and have folks test! What do you think?

That would be great!

**Comment by msalsbery** [comment]
@iangilman On #1863 we agreed to get that one in the 2.5.0 milestone as well. Not necessary, as the majority of the old IE support was probably in MouseTracker and is now gone, but I can make a pass on the rest of the code to finish #1863 before we release a 2.5.0...

Would be easier to do if this one gets merged first ;-)

**Comment by iangilman** [comment]
Awesome… Merging! @Msalsbery, thank you for taking on this Herculean task!

As for finishing up #1863, I'd certainly be happy for it if you wanted to. I figure we should give this patch some testing time on master before we release it, so there is certainly time for other stuff in the meantime.

**Comment by iangilman** [comment]
@msalsbery, would you mind going through and closing all of the issues fixed by this patch? :-)

**Comment by msalsbery** [comment]
> Awesome… Merging! @msalsbery, thank you for taking on this Herculean task!
> 
> As for finishing up #1863, I'd certainly be happy for it if you wanted to. I figure we should give this patch some testing time on master before we release it, so there is certainly time for other stuff in the meantime.

Yay thanks!

For #1863 there's probably not a whole lot to do...a lot of the legacy support was in MouseTracker

**Comment by msalsbery** [comment]
> @msalsbery, would you mind going through and closing all of the issues fixed by this patch? :-)

Will do thanks

**Comment by msalsbery** [comment]
> Now that this patch has landed, our master build is failing some of its unit tests:
> 
> https://travis-ci.org/github/openseadragon/openseadragon/builds/761987114
> 
> Seems odd, since this branch on its own passed the tests just fine. I've already restarted the build once, thinking it might be an intermittent error, but failed the second time too. Can you take a look? I can try restarting the build again if you'd like…

Ugh! While removing old IE code I found the legacy mouse shim used by the tests needed updating so I may have just fixed it already. Is weird though, as I did run tests...just not after last few commits darnit.

Looking into it...

**Pull request merged**
